Source/WebCore/Modules/webcodecs/WebCodecsBase.h

```diff
@@ -0,0 +1,102 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(WEB_CODECS)
+
+#include "ActiveDOMObject.h"
+#include "EventTarget.h"
+#include "WebCodecsCodecState.h"
+#include <wtf/Deque.h>
+#include <wtf/TZoneMalloc.h>
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WebCore {
+
+class WebCodecsControlMessage;
+
+// WebCodecsBase implements the "Control Message Queue"
+// as per https://w3c.github.io/webcodecs/#control-message-queue-slot
+// And handle "Codec Saturation"
+// as per https://w3c.github.io/webcodecs/#saturated
+class WebCodecsBase
+    : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<WebCodecsBase>
+    , public ActiveDOMObject
+    , public EventTarget {
+    WTF_MAKE_TZONE_OR_ISO_ALLOCATED(WebCodecsBase);
+public:
+    virtual ~WebCodecsBase();
+
+    WebCodecsCodecState state() const { return m_state; }
+
+    // ActiveDOMObject.
+    void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::ref(); }
+    void deref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::deref(); }
+    bool virtualHasPendingActivity() const final;
+
+protected:
+    WebCodecsBase(ScriptExecutionContext&);
+    ScriptExecutionContext* scriptExecutionContext() const final { return ActiveDOMObject::scriptExecutionContext(); }
+
+    void setState(WebCodecsCodecState state) { m_state = state; }
+
+    size_t operationQueueSize() const { return m_operationQueueSize; }
+    // Equivalent to spec's "Increment [[encodeQueueSize]]." or "Increment [[decodeQueueSize]]"
+    void incrementOperationQueueSize();
+    // Equivalent to spec's "Decrement [[encodeQueueSize]] or "Decrement [[decodeQueueSize]]" and run the Schedule Dequeue Event algorithm"
+    void decrementOperationQueueSizeAndScheduleDequeueEvent();
+
+    void queueControlMessageAndProcess(WebCodecsControlMessage&&);
+    void queueControlMessageForCodecOperationAndProcess(WebCodecsControlMessage&&);
+    void processControlMessageQueue();
+    void clearControlMessageQueue();
+    void clearControlMessageQueueAndMaybeScheduleDequeueEvent();
+    void blockControlMessageQueue();
+    void unblockControlMessageQueue();
+
+    virtual size_t maximumCodecOperationsEnqueued() const { return 1; }
+    void increaseCodecOperationCount() { m_codecOperationsEnqueued++; };
+    void decreaseCodecOperationCountAndMaybeProcessControlMessageQueue();
+
+private:
+    // EventTarget
+    void refEventTarget() final { ref(); }
+    void derefEventTarget() final { deref(); }
+
+    bool isCodecSaturated() const { return m_codecOperationsEnqueued >= maximumCodecOperationsEnqueued(); }
+    void scheduleDequeueEvent();
+
+    bool m_isMessageQueueBlocked = false;
+    size_t m_operationQueueSize { 0 };
+    size_t m_codecOperationsEnqueued { 0 };
```

Comment:
This is not very clear as a name.
Can we name it to something like `m_pendingEncodingOrDecodingCount`?